lib/benchmark/bench.go

```diff
@@ -59,175 +57,169 @@ type Result struct {
 	RequestsOriginated int
 	// RequestsFailed is amount of requests failed
 	RequestsFailed int
-	// Histogram is a duration histogram
-	Histogram *hdrhistogram.Histogram
+	// ResponseHistogram is a duration actual request histogram
+	ResponseHistogram *hdrhistogram.Histogram
+	//ServiceHistogram is the duration of service histogram
+	ServiceHistogram *hdrhistogram.Histogram
 	// LastError contains last recorded error
 	LastError error
 	// Duration it takes for the whole benchmark to run
 	Duration time.Duration
 }
 
+//
+
 // Benchmark connects to remote server and executes requests in parallel according
 // to benchmark spec. It returns benchmark result when completed.
 // This is a blocking function that can be cancelled via context argument.
 func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Result, error) {
 	tc.Stdout = ioutil.Discard
 	tc.Stderr = ioutil.Discard
 	tc.Stdin = &bytes.Buffer{}
-	workerCtx, cancelWorkers := context.WithCancel(context.Background())
-	defer cancelWorkers()
-
-	requestC := make(chan benchMeasure)
-	responseC := make(chan benchMeasure, c.Threads)
-
-	for i := 0; i < c.Threads; i++ {
-		thread := &benchmarkThread{
-			id:          i,
-			ctx:         ctx,
-			client:      tc,
-			command:     c.Command,
-			interactive: c.Interactive,
-			receiveC:    requestC,
-			sendC:       responseC,
+	var delay time.Duration = 0
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	requestsC := make(chan *benchMeasure)
+	resultC := make(chan *benchMeasure)
+
+	go func() {
+		interval := time.Duration(float64(1) / float64(c.Rate) * float64(time.Second))
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		start := time.Now()
+		for {
+			select {
+			case <-ticker.C:
+				delay = delay + interval
+				t := start.Add(delay)
```

Comment:
I think a comment about why you choose `t := start.Add(delay)` wouldn't go awry here.  Particularly getting at why this is preferred to `t := time.Now()`.  It may be obvious to those who have thought about coordinated omission, but I'd guess it isn't intuitive to folks new to the codebase.